sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/operator/StringOperators.java

```diff
@@ -0,0 +1,242 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.sdk.extensions.sql.impl.interpreter.operator;
+
+import java.util.List;
+import org.apache.calcite.runtime.SqlFunctions;
+import org.apache.calcite.sql.fun.SqlTrimFunction;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+/** String operator implementations. */
+public class StringOperators {
+
+  /** A {@link BeamSqlOperator} that returns a string. */
+  public interface StringOperator extends BeamSqlOperator {
+    default SqlTypeName getOutputType() {
+      return SqlTypeName.VARCHAR;
+    }
+  }
+
+  @FunctionalInterface
+  private interface StringUnaryOperator extends BeamSqlUnaryOperator {
+    default boolean accept(BeamSqlExpression arg) {
+      return SqlTypeName.CHAR_TYPES.contains(arg.getOutputType());
+    }
+
+    default SqlTypeName getOutputType() {
+      return SqlTypeName.VARCHAR;
+    }
+  }
+
+  public static final BeamSqlOperator CHAR_LENGTH =
+      (StringUnaryOperator)
+          (BeamSqlPrimitive arg) ->
+              BeamSqlPrimitive.of(SqlTypeName.INTEGER, SqlFunctions.charLength(arg.getString()));
+
+  public static final BeamSqlOperator UPPER =
+      (StringUnaryOperator)
+          (BeamSqlPrimitive arg) ->
+              BeamSqlPrimitive.of(SqlTypeName.VARCHAR, SqlFunctions.upper(arg.getString()));
+
+  public static final BeamSqlOperator LOWER =
+      (StringUnaryOperator)
+          (BeamSqlPrimitive arg) ->
+              BeamSqlPrimitive.of(SqlTypeName.VARCHAR, SqlFunctions.lower(arg.getString()));
+
+  /** {@code INITCAP}. */
+  public static final BeamSqlOperator INIT_CAP =
+      (StringUnaryOperator)
+          (BeamSqlPrimitive arg) ->
+              BeamSqlPrimitive.of(SqlTypeName.VARCHAR, SqlFunctions.initcap(arg.getString()));
```

Comment:
Can these be put into maps? E.g.:

```
UNARY_FUNCTIONS = ImmutableMap.<SqlOperator, Function<String, String>>of(
   SqlStdOperatorTable.INITCAP, SqlFunctions::initcap,
   ...
);

BINARY_FUNCTIONS = ImmutableMap.<SqlOperator, BiFunction<String, String, String>>of(

```

Reply:
It isn't quite so straightforward - the `SqlFunctions` need wrappers in a lot of cases. Also I need to double-check that we don't need `accept` and then remove it too. I think the most useful map would be to replace the giant `switch` in `BeamSqlFnExecutor`, right? I think we should definitely get to that point, where we just have one case for mapping any operator to a `BeamSqlOperatorExpression`.